lib/ewasm/contract.js

```diff
@@ -0,0 +1,32 @@
+const Env = require('./env')
+const Memory = require('./memory')
+const { VmError } = require('../exceptions')
+
+module.exports = class Contract {
+  constructor (code) {
+    this._module = new WebAssembly.Module(code)
+  }
+
+  run (opts) {
+    const env = new Env(opts)
+
+    this._instance = new WebAssembly.Instance(this._module, env.imports)
+    if (!this._instance.exports.main || !this._instance.exports.memory) {
+      throw new Error('EWASM module has no main or memory export')
+    }
+
+    this._memory = new Memory(this._instance.exports.memory)
+    env.setMemory(this._memory)
+
+    // Run contract. It throws even on successful finish.
+    try {
+      this._instance.exports.main()
+    } catch (e) {
+      if (!(e instanceof VmError) && e.message !== 'unreachable') {
+        throw e
```

Comment:
Why not rethrow `VmError`?

I think what this should do is only catch a single exception: `FinishExecution`, which is emitted in `finish` (just create a new exception type). Every other exception is real and should not be ignored.

Reply:
Well you could wrap wasm execution errors into a `VmError` exception if we want to hide wasm exceptions - we may already do that for `opFns` in `runCode`, wrapping them into "internal errors".

Reply:
I used `VmError` to remain consistent with `opFns`'s error handling, and it's thrown for out of gas and revert. The reason why I catch this here is because `runCode` doesn't do any error handling, it excepts the error to be in the `results` object. Maybe this is something that can be changed in refactoring?
Is it always the case that ewasm will throw `unreachable` immediately after calling `finish`? In that case yeah it might be better to throw a custom error in `finish`.

Reply:
There's no requirement to have `unreachable` after `finish` - the finish method should abort execution.

Rust inserts the unreachable opcode after functions which are marked non-returned and `finish` is marked as such in ewasm-rust-api, but this code here should not rely on that.